packages/core/src/actions.ts

```diff
@@ -582,107 +582,142 @@ export function resolveActions<TContext, TEvent extends EventObject>(
+            >).get(updatedContext, _event.data);
+            if (!matchedActions) {
+              return [];
+            }
+            const [resolvedActionsFromPure, resolvedContext] = resolveActions(
+              machine,
+              currentState,
+              updatedContext,
+              _event,
+              toActionObjects(toArray(matchedActions), machine.options.actions)
             );
-          })?.actions;
-
-          if (!matchedActions) {
-            return [];
+            updatedContext = resolvedContext;
+            preservedContexts?.push(updatedContext);
+            return resolvedActionsFromPure;
           }
-
-          const resolved = resolveActions(
-            machine,
-            currentState,
-            updatedContext,
-            _event,
-            toActionObjects(toArray(matchedActions), machine.options.actions)
-          );
-          updatedContext = resolved[1];
-          return resolved[0];
-        }
-        case actionTypes.pure: {
-          const matchedActions = (actionObject as PureAction<
-            TContext,
-            TEvent
-          >).get(updatedContext, _event.data);
-          if (!matchedActions) {
-            return [];
+          case actionTypes.stop: {
+            return resolveStop(
+              actionObject as StopAction<TContext, TEvent>,
+              updatedContext,
+              _event
+            );
           }
-          const resolved = resolveActions(
-            machine,
-            currentState,
-            updatedContext,
-            _event,
-            toActionObjects(toArray(matchedActions), machine.options.actions)
-          );
-          updatedContext = resolved[1];
-          return resolved[0];
-        }
-        case actionTypes.stop: {
-          return resolveStop(
-            actionObject as StopAction<TContext, TEvent>,
-            updatedContext,
-            _event
-          );
+          case actionTypes.assign: {
+            updatedContext = updateContext(
+              updatedContext,
+              _event,
+              [actionObject as AssignAction<TContext, TEvent>],
+              currentState
+            );
+            preservedContexts?.push(updatedContext);
+            break;
+          }
+          default:
+            const resolvedActionObject = toActionObject(
+              actionObject,
+              machine.options.actions
+            );
+            const { exec } = resolvedActionObject;
+            if (exec && preservedContexts) {
+              const contextIndex = preservedContexts.length - 1;
+              resolvedActionObject.exec = (_ctx, ...args) => {
+                exec?.(preservedContexts[contextIndex], ...args);
+              };
+            }
+            return resolvedActionObject;
         }
-        default:
-          return toActionObject(actionObject, machine.options.actions);
-      }
-    })
+      })
+      .filter((a): a is ActionObject<TContext, TEvent> => !!a)
```

Comment:
What has changed that this `filter` is now needed? We were not filtering before those changes

Reply:
`actionTypes.assign` returns `undefined`, not an `ActionObject`